### PR TITLE
Features/ilu 242 UI validation

### PR DESCRIPTION
**Pull request by thatisdrtruman**
Tests and UI test fully fixed. Ready to merge

### CODE REVIEW

File: ips/__init__.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import logging
+from sys import exc_info
```
Comment (bitmonkey) [inline ips/__init__.py]: You don't need this?

File: ips/services/forms.py [hunk 2]
```diff
@@ -116,15 +117,40 @@ class ImportPVForm(FlaskForm):
...
+    air_file = FileField(validators=[FileAllowed(['csv'], 'File must be a .csv file.')])
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        fieldnames = [
+            'survey_file', 'shift_file', 'non_response_file', 'unsampled_file',
+            'tunnel_file', 'sea_file', 'air_file' 
+        ]
+
```
Comment (bitmonkey) [inline ips/services/forms.py:135]: Would be useful to leave a comment here explaining what is happening i.e that you're looping through the fields adding a FileRequired validator to the _extra_validator attribute for fields that haven't had a filename set in the meta object.

File: ips/services/forms.py [hunk 3]
```diff
@@ -116,15 +117,40 @@ class ImportPVForm(FlaskForm):
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        fieldnames = [
+            'survey_file', 'shift_file', 'non_response_file', 'unsampled_file',
+            'tunnel_file', 'sea_file', 'air_file' 
+        ]
+    
+        validators = {}
```
Comment (bitmonkey) [inline ips/services/forms.py:136]: Would be useful to leave a comment here explaining what is happening i.e that you're looping through the fields adding a FileRequired validator to the _extra_validator attribute for fields that haven't had a filename set in the meta object.

File: ips/services/new_run_steps/step_5.py [hunk 4]
```diff
@@ -1,79 +1,104 @@
...
+from flask_wtf.file import FileRequired
 
 from ips.services import app_methods
 from ips.services.forms import LoadDataForm
 
-
 def run_step_5(run_id):
-    form = LoadDataForm()
+    from remote_pdb import RemotePdb
+    #RemotePdb('0.0.0.0', 4445).set_trace()
```
Comment (bitmonkey) [inline ips/services/new_run_steps/step_5.py:10]: Don't check in unneeded / commented out code.

File: ips/services/forms.py (hunk 3 above)
Comment (bitmonkey) [inline ips/services/forms.py:136]: Would be useful to leave a comment here explaining what is happening i.e that you're looping through the fields adding a FileRequired validator to the _extra_validator attribute for fields that haven't had a filename set in the meta object.

File: ips/__init__.py (hunk 1 above)
Comment (bitmonkey) [inline ips/__init__.py]: This isn't needed....you're not using exc_info anywhere

Comment (thatisdrtruman) [reply ips/__init__.py]: Sure, fixing it now

File: ips/services/new_run_steps/step_5.py [hunk 5]
```diff
@@ -1,79 +1,104 @@
...
-    form = LoadDataForm()
+    from remote_pdb import RemotePdb
+    #RemotePdb('0.0.0.0', 4445).set_trace() 
     error = False
-
-    if form.validate_on_submit():
-
+    run_info = app_methods.get_run(run_id)
+        
+    #RemotePdb('0.0.0.0', 4445).set_trace()
```
Comment (bitmonkey) [inline ips/services/new_run_steps/step_5.py:14]: More commented out code (and un-needed as just for debugging)

File: ips/services/new_run_steps/step_5.py [hunk 6]
```diff
@@ -1,79 +1,104 @@
...
+        'survey_file': run_info['SURVEY_FILE'],
+        'shift_file': run_info['SHIFT_FILE'],
+        'non_response_file': run_info['NR_FILE'],
+        'unsampled_file': run_info['UNSAMPLED_FILE'],
+        'tunnel_file': run_info['TUNNEL_FILE'],
+        'sea_file': run_info['SEA_FILE'],
+        'air_file': run_info['AIR_FILE']
+    }
+    form = LoadDataForm(meta=meta)
+    #RemotePdb('0.0.0.0', 4445).set_trace()
```
Comment (bitmonkey) [inline ips/services/new_run_steps/step_5.py:27]: Another commented out piece of debugging code.

**Review by thatisdrtruman — COMMENTED** [review]
That should be cleaner now

File: tests/test_step_5.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+                assert field not in loadform.errors
+            else:
+                assert field in loadform._extra_validators, f"Extra validator expected if {field} filename supplied"
+                assert any([
+                    isinstance(validator, FileRequired) for validator in loadform._extra_validators[field]
+                ]) is True, f"FileRequired validator expected as extra validator when no filename supplied for {field}"
+                assert loadform.errors[field] == ['This field is required.'], f"Expected {field} field validation to fail as no filename givened"
+    client.post("/")
+
+def test_dynamic_meeting(app, client):
```
Comment (bitmonkey) [inline tests/test_step_5.py:102]: This looks like an empty test? Guessing we can remove?

**Review by bitmonkey — CHANGES_REQUESTED** [review]

**Pull request closed without merging**